File: frontend/src/loan-scoring-demo/scoringRules.ts
import type {
  BorrowerComponentKey,
  BorrowerScoreRecord,
  LoanScoringInput,
  ScoreComponent,
  ScoreComponentDefinition,
  TransactionComponentKey,
} from "./types";
import { clamp, daysBetweenIso, hasNonEmptyText, isPositiveFiniteNumber, normalizeToken, round2 } from "./utils";

export const BORROWER_COMPONENT_DEFINITIONS: ScoreComponentDefinition<BorrowerComponentKey>[] = [
  {
    key: "repayment_history",
    label: "Past repayment history on our platform",
    weight_percent: 20,
  },
  {
    key: "current_platform_exposure",
    label: "Current exposure on our platform",
    weight_percent: 10,
  },
  {
    key: "business_age_continuity",
    label: "Business age and operating continuity",
    weight_percent: 5,
  },
  {
    key: "legal_compliance",
    label: "Legal/compliance status",
    weight_percent: 4,
  },
  {
    key: "kyb_consistency",
    label: "Basic borrower identity and KYB consistency",
    weight_percent: 2,
  },
  {
    key: "cdi_activity_continuity",
    label: "CDI-observed business activity continuity",
    weight_percent: 4,
  },
];

export const TRANSACTION_COMPONENT_DEFINITIONS: ScoreComponentDefinition<TransactionComponentKey>[] = [
  {
    key: "loan_to_invoice_reasonableness",
    label: "Loan-to-invoice / advance-rate reasonableness",
    weight_percent: 8,
  },
  {
    key: "collateral_recovery_quality",
    label: "Collateral/recovery quality",
    weight_percent: 10,
  },
  {
    key: "platform_order_normality",
    label: "Order normality based on our platform history",
    weight_percent: 7,
  },
  {
    key: "basic_transaction_completeness",
    label: "Basic transaction completeness",
    weight_percent: 5,
  },
  {
    key: "verified_trade_shipment",
    label: "Verified trade/shipment data",
    weight_percent: 14,
  },
  {
    key: "supplier_reliability",
    label: "Supplier reliability",
    weight_percent: 4,
  },
  {
    key: "cdi_order_normality",
    label: "CDI-based order normality",
    weight_percent: 4,
  },
  {
    key: "buyer_channel_concentration",
    label: "Buyer/customer/channel concentration risk",
    weight_percent: 3,
  },
];

export function buildScoreComponent<Key extends string>(
  definition: ScoreComponentDefinition<Key>,
  rawScore: number,
): ScoreComponent {
  const raw_score_0_100 = round2(clamp(rawScore, 0, 100));

  return {
    key: definition.key,
    label: definition.label,
    weight_percent: definition.weight_percent,
    raw_score_0_100,
    weighted_points: round2((raw_score_0_100 * definition.weight_percent) / 100),
  };
}

export function buildBorrowerScoreComponents(record: BorrowerScoreRecord): ScoreComponent[] {
  return BORROWER_COMPONENT_DEFINITIONS.map((definition) =>
    buildScoreComponent(definition, record.component_scores[definition.key]),
  );
}

export function getInternalShowstopper(input: LoanScoringInput): string | null {
  if (input.loan_request.loan_amount_hkd <= 0) {
    return "Loan amount must be positive.";
  }

  if (input.loan_request.loan_duration_days <= 0) {
    return "Loan duration must be positive.";
  }

  if (input.transaction.invoice.invoice_value_hkd <= 0) {
    return "Invoice value must be positive.";
  }

  if (input.loan_request.loan_amount_hkd > input.transaction.invoice.invoice_value_hkd) {
    return "Requested loan amount exceeds invoice value.";
  }

  return null;
}

export function buildLocalTransactionComponents(
  input: LoanScoringInput,
  borrowerRecord: BorrowerScoreRecord,
): {
  components: Partial<Record<TransactionComponentKey, ScoreComponent>>;
  showstopper: string | null;
} {
  const loanToInvoice = scoreLoanToInvoiceReasonableness(input);

  return {
    components: {
      loan_to_invoice_reasonableness: buildScoreComponent(
        getTransactionDefinition("loan_to_invoice_reasonableness"),
        loanToInvoice.rawScore,
      ),
      collateral_recovery_quality: buildScoreComponent(
        getTransactionDefinition("collateral_recovery_quality"),
        scoreCollateralRecoveryQuality(input),
      ),
      platform_order_normality: buildScoreComponent(
        getTransactionDefinition("platform_order_normality"),
        scorePlatformOrderNormality(input, borrowerRecord),
      ),
      basic_transaction_completeness: buildScoreComponent(
        getTransactionDefinition("basic_transaction_completeness"),
        scoreBasicTransactionCompleteness(input),
      ),
    },
    showstopper: loanToInvoice.showstopper,
  };
}

export function sumWeightedPoints(components: ScoreComponent[]): number {
  return round2(components.reduce((sum, component) => sum + component.weighted_points, 0));
}

export function getTransactionDefinition(
  key: TransactionComponentKey,
): ScoreComponentDefinition<TransactionComponentKey> {
  const definition = TRANSACTION_COMPONENT_DEFINITIONS.find((candidate) => candidate.key === key);

  if (!definition) {
    throw new Error(`Unknown transaction score component: ${key}`);
  }

  return definition;
}

function scoreLoanToInvoiceReasonableness(input: LoanScoringInput): {
  rawScore: number;
  showstopper: string | null;
} {
  const loanAmount = input.loan_request.loan_amount_hkd;
  const invoiceValue = input.transaction.invoice.invoice_value_hkd;

  if (invoiceValue <= 0 || loanAmount <= 0) {
    return { rawScore: 0, showstopper: null };
  }

  const advanceRate = loanAmount / invoiceValue;

  if (advanceRate <= 0.6) {
    return { rawScore: 95, showstopper: null };
  }

  if (advanceRate <= 0.75) {
    return { rawScore: 85, showstopper: null };
  }

  if (advanceRate <= 0.9) {
    return { rawScore: 65, showstopper: null };
  }

  if (advanceRate <= 1) {
    return { rawScore: 40, showstopper: null };
  }

  return {
    rawScore: 10,
    showstopper: "Requested loan amount exceeds invoice value.",
  };
}

function scoreCollateralRecoveryQuality(input: LoanScoringInput): number {
  const collateral = input.transaction.collateral;
  const loanAmount = input.loan_request.loan_amount_hkd;
  const baseByType = {
    cash_deposit: 95,
    warehouse_receipt: 85,
    insured_goods: 75,
    inventory: 55,
    none: 25,
  } satisfies Record<LoanScoringInput["transaction"]["collateral"]["type"], number>;
  const baseScore = baseByType[collateral.type];
  const coverage = loanAmount > 0 ? Math.max(0, collateral.value_hkd) / loanAmount : 0;

  if (coverage >= 1) {
    return baseScore;
  }

  if (coverage >= 0.75) {
    return clamp(baseScore - 5, 0, 100);
  }

  if (coverage >= 0.5) {
    return clamp(baseScore - 15, 0, 100);
  }

  if (coverage > 0) {
    return clamp(baseScore - 30, 0, 100);
  }

  return Math.min(baseScore, 20);
}

function scorePlatformOrderNormality(
  input: LoanScoringInput,
  borrowerRecord: BorrowerScoreRecord,
): number {
  const history = borrowerRecord.platform_history;

  if (!history) {
    return 45;
  }

  let score = 100;
  const invoiceValue = input.transaction.invoice.invoice_value_hkd;
  const medianInvoiceValue = history.median_invoice_value_hkd;
  const loanDuration = input.loan_request.loan_duration_days;
  const usualDuration = history.usual_loan_duration_days;
  const productType = normalizeToken(input.transaction.goods.product_type);
  const supplierCountry = input.transaction.supplier.supplier_country.trim().toUpperCase();

  if (medianInvoiceValue > 0) {
    if (invoiceValue > medianInvoiceValue * 2.5) {
      score -= 35;
    } else if (invoiceValue > medianInvoiceValue * 1.5) {
      score -= 15;
    }
  }

  if (usualDuration > 0) {
    if (loanDuration > usualDuration * 1.75) {
      score -= 20;
    } else if (loanDuration > usualDuration * 1.25) {
      score -= 10;
    }
  }

  const commonProductTypes = history.common_product_types.map(normalizeToken);
  const commonSupplierCountries = history.common_supplier_countries.map((country) =>
    country.trim().toUpperCase(),
  );

  if (!commonProductTypes.includes(productType)) {
    score -= 15;
  }

  if (!commonSupplierCountries.includes(supplierCountry)) {
    score -= 10;
  }

  return clamp(score, 0, 100);
}

function scoreBasicTransactionCompleteness(input: LoanScoringInput): number {
  const shipmentDays = daysBetweenIso(
    input.transaction.shipment.expected_departure_date,
    input.transaction.shipment.expected_arrival_date,
  );
  const checks = [
    isPositiveFiniteNumber(input.transaction.purchase_order.po_value_hkd),
    isPositiveFiniteNumber(input.transaction.invoice.invoice_value_hkd),
    hasNonEmptyText(input.transaction.supplier.supplier_name),
    hasNonEmptyText(input.transaction.supplier.supplier_country),
    hasNonEmptyText(input.transaction.goods.product_type),
    Number.isFinite(shipmentDays) && shipmentDays > 0,
    hasNonEmptyText(input.loan_request.expected_repayment_source),
    isPositiveFiniteNumber(input.loan_request.loan_amount_hkd),
    isPositiveFiniteNumber(input.loan_request.loan_duration_days),
  ];
  const invalidCount = checks.filter((isValid) => !isValid).length;

  return clamp(100 - invalidCount * 10, 0, 100);
}

File: backend/src/services/loan_scoring/scoring_rules.py
from .types import (
    BorrowerScoreRecord,
    LoanScoringInput,
    ScoreComponent,
    ScoreComponentDefinition,
)
from .utils import clamp, days_between_iso, has_text, normalize_token, positive_number, round2

BORROWER_COMPONENT_DEFINITIONS = [
    ScoreComponentDefinition(
        key="repayment_history",
        label="Past repayment history on our platform",
        weight_percent=20,
    ),
    ScoreComponentDefinition(
        key="current_platform_exposure",
        label="Current exposure on our platform",
        weight_percent=10,
    ),
    ScoreComponentDefinition(
        key="business_age_continuity",
        label="Business age and operating continuity",
        weight_percent=5,
    ),
    ScoreComponentDefinition(
        key="legal_compliance",
        label="Legal/compliance status",
        weight_percent=4,
    ),
    ScoreComponentDefinition(
        key="kyb_consistency",
        label="Basic borrower identity and KYB consistency",
        weight_percent=2,
    ),
    ScoreComponentDefinition(
        key="cdi_activity_continuity",
        label="CDI-observed business activity continuity",
        weight_percent=4,
    ),
]

TRANSACTION_COMPONENT_DEFINITIONS = [
    ScoreComponentDefinition(
        key="loan_to_invoice_reasonableness",
        label="Loan-to-invoice / advance-rate reasonableness",
        weight_percent=8,
    ),
    ScoreComponentDefinition(
        key="collateral_recovery_quality",
        label="Collateral/recovery quality",
        weight_percent=10,
    ),
    ScoreComponentDefinition(
        key="platform_order_normality",
        label="Order normality based on our platform history",
        weight_percent=7,
    ),
    ScoreComponentDefinition(
        key="basic_transaction_completeness",
        label="Basic transaction completeness",
        weight_percent=5,
    ),
    ScoreComponentDefinition(
        key="verified_trade_shipment",
        label="Verified trade/shipment data",
        weight_percent=14,
    ),
    ScoreComponentDefinition(
        key="supplier_reliability",
        label="Supplier reliability",
        weight_percent=4,
    ),
    ScoreComponentDefinition(
        key="cdi_order_normality",
        label="CDI-based order normality",
        weight_percent=4,
    ),
    ScoreComponentDefinition(
        key="buyer_channel_concentration",
        label="Buyer/customer/channel concentration risk",
        weight_percent=3,
    ),
]


def build_score_component(
    definition: ScoreComponentDefinition,
    raw_score: float,
) -> ScoreComponent:
    raw_score_0_100 = round2(clamp(raw_score, 0, 100))
    return ScoreComponent(
        key=definition.key,
        label=definition.label,
        weight_percent=definition.weight_percent,
        raw_score_0_100=raw_score_0_100,
        weighted_points=round2(raw_score_0_100 * definition.weight_percent / 100),
    )


def build_borrower_score_components(record: BorrowerScoreRecord) -> list[ScoreComponent]:
    component_scores = record.component_scores.model_dump()
    return [
        build_score_component(definition, component_scores[definition.key])
        for definition in BORROWER_COMPONENT_DEFINITIONS
    ]


def get_internal_showstopper(input_data: LoanScoringInput) -> str | None:
    if input_data.loan_request.loan_amount_hkd <= 0:
        return "Loan amount must be positive."
    if input_data.loan_request.loan_duration_days <= 0:
        return "Loan duration must be positive."
    if input_data.transaction.invoice.invoice_value_hkd <= 0:
        return "Invoice value must be positive."
    if input_data.loan_request.loan_amount_hkd > input_data.transaction.invoice.invoice_value_hkd:
        return "Requested loan amount exceeds invoice value."
    return None


def build_local_transaction_components(
    input_data: LoanScoringInput,
    borrower_record: BorrowerScoreRecord,
) -> tuple[dict[str, ScoreComponent], str | None]:
    raw_score, showstopper = _score_loan_to_invoice_reasonableness(input_data)
    return (
        {
            "loan_to_invoice_reasonableness": build_score_component(
                _transaction_definition("loan_to_invoice_reasonableness"),
                raw_score,
            ),
            "collateral_recovery_quality": build_score_component(
                _transaction_definition("collateral_recovery_quality"),
                _score_collateral_recovery_quality(input_data),
            ),
            "platform_order_normality": build_score_component(
                _transaction_definition("platform_order_normality"),
                _score_platform_order_normality(input_data, borrower_record),
            ),
            "basic_transaction_completeness": build_score_component(
                _transaction_definition("basic_transaction_completeness"),
                _score_basic_transaction_completeness(input_data),
            ),
        },
        showstopper,
    )


def sum_weighted_points(components: list[ScoreComponent]) -> float:
    return round2(sum(component.weighted_points for component in components))


def transaction_definition(key: str) -> ScoreComponentDefinition:
    return _transaction_definition(key)


def _transaction_definition(key: str) -> ScoreComponentDefinition:
    for definition in TRANSACTION_COMPONENT_DEFINITIONS:
        if definition.key == key:
            return definition
    raise ValueError(f"Unknown transaction score component: {key}")


def _score_loan_to_invoice_reasonableness(
    input_data: LoanScoringInput,
) -> tuple[float, str | None]:
    loan_amount = input_data.loan_request.loan_amount_hkd
    invoice_value = input_data.transaction.invoice.invoice_value_hkd
    if invoice_value <= 0 or loan_amount <= 0:
        return 0, None

    advance_rate = loan_amount / invoice_value
    if advance_rate <= 0.60:
        return 95, None
    if advance_rate <= 0.75:
        return 85, None
    if advance_rate <= 0.90:
        return 65, None
    if advance_rate <= 1.00:
        return 40, None
    return 10, "Requested loan amount exceeds invoice value."


def _score_collateral_recovery_quality(input_data: LoanScoringInput) -> float:
    collateral = input_data.transaction.collateral
    loan_amount = input_data.loan_request.loan_amount_hkd
    base_by_type = {
        "cash_deposit": 95,
        "warehouse_receipt": 85,
        "insured_goods": 75,
        "inventory": 55,
        "none": 25,
    }
    base_score = base_by_type[collateral.type]
    coverage = max(0, collateral.value_hkd) / loan_amount if loan_amount > 0 else 0

    if coverage >= 1:
        return base_score
    if coverage >= 0.75:
        return clamp(base_score - 5, 0, 100)
    if coverage >= 0.5:
        return clamp(base_score - 15, 0, 100)
    if coverage > 0:
        return clamp(base_score - 30, 0, 100)
    return min(base_score, 20)


def _score_platform_order_normality(
    input_data: LoanScoringInput,
    borrower_record: BorrowerScoreRecord,
) -> float:
    history = borrower_record.platform_history
    if not history:
        return 45

    score = 100
    invoice_value = input_data.transaction.invoice.invoice_value_hkd
    if history.median_invoice_value_hkd > 0:
        if invoice_value > history.median_invoice_value_hkd * 2.5:
            score -= 35
        elif invoice_value > history.median_invoice_value_hkd * 1.5:
            score -= 15

    loan_duration = input_data.loan_request.loan_duration_days
    if history.usual_loan_duration_days > 0:
        if loan_duration > history.usual_loan_duration_days * 1.75:
            score -= 20
        elif loan_duration > history.usual_loan_duration_days * 1.25:
            score -= 10

    product_type = normalize_token(input_data.transaction.goods.product_type)
    supplier_country = input_data.transaction.supplier.supplier_country.strip().upper()
    common_product_types = [normalize_token(value) for value in history.common_product_types]
    common_supplier_countries = [
        value.strip().upper() for value in history.common_supplier_countries
    ]

    if product_type not in common_product_types:
        score -= 15
    if supplier_country not in common_supplier_countries:
        score -= 10
    return clamp(score, 0, 100)


def _score_basic_transaction_completeness(input_data: LoanScoringInput) -> float:
    shipment_days = days_between_iso(
        input_data.transaction.shipment.expected_departure_date,
        input_data.transaction.shipment.expected_arrival_date,
    )
    checks = [
        positive_number(input_data.transaction.purchase_order.po_value_hkd),
        positive_number(input_data.transaction.invoice.invoice_value_hkd),
        has_text(input_data.transaction.supplier.supplier_name),
        has_text(input_data.transaction.supplier.supplier_country),
        has_text(input_data.transaction.goods.product_type),
        shipment_days > 0,
        has_text(input_data.loan_request.expected_repayment_source),
        positive_number(input_data.loan_request.loan_amount_hkd),
        positive_number(input_data.loan_request.loan_duration_days),
    ]
    return clamp(100 - checks.count(False) * 10, 0, 100)
